jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java

```diff
@@ -1279,13 +1328,42 @@ else if (_path != null)
          */
         private void checkSegment(String uri, int segment, int end, boolean param)
         {
-            if (!_ambiguous.contains(Ambiguous.SEGMENT))
+            // We had a non last empty segment meaning it was ambiguous.
+            if (_emptySegment)
+                _ambiguous.add(Ambiguous.EMPTY);
+
+            if (end == segment)
+            {
+                // Empty segments are only ambiguous if followed by a '#', '?' or end of string.
+                if (end >= uri.length() || ("#?".indexOf(uri.charAt(end)) >= 0))
+                    return;
+
+                // If the first segment is empty then it is ambiguous.
+                if (segment == 0)
+                {
+                    _ambiguous.add(Ambiguous.EMPTY);
+                    return;
+                }
+
+                // Otherwise only a non last empty segment is ambiguous.
+                if (!_emptySegment)
+                {
+                    _emptySegment = true;
```

Comment:
I fail to understand why you need to remember `_emptySegment` across calls.
Here, can't you just call `_ambiguous.add(Ambiguous.EMPTY)` and get rid of the `_emptySegment` field?

Seems to me that `/a//p;/a?#` would add an empty segment multiple times -- sure it's a `Set` but I don't understand why we would do that?

If it is necessary, needs a comment.

Reply:
@sbordet There are comments explaining:
```java
            // We had a non last empty segment meaning it was ambiguous.
...
                // Empty segments are only ambiguous if followed by a '#', '?' or end of string.
...
                // If the first segment is empty then it is ambiguous.
...
                // Otherwise only a non last empty segment is ambiguous.
```

The `_emptySegment` field is true if there has been an empty segment, but that is not necessarily ambiguous.  It is only ambiguous if there is another segment found after an empty segment.  Hence the `_emptySegment` check at the start of the method.

Ah but there is a mistake in one of the comments.  I will comment separately